enhancements/cluster-logging/loki-storage.md

```diff
@@ -0,0 +1,623 @@
+
+    // ManagementState defines if the CR should be managed by the operator or not.
+    // Default is managed.
+    //
+    // +required
+    // +kubebuilder:default:=Managed
+    // +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Management State"
+    ManagementState ManagementStateType `json:"managementState,omitempty"`
+
+    // Size defines one of the support Loki deployment scale out sizes.
+    //
+    // +required
+    Size LokiStackSizeType `json:"size,omitempty"`
+
+    // Storage defines the spec for the object storage endpoint to store logs.
+    //
+    // +required
+    Storage ObjectStorageSpec `json:"storage,omitempty"`
+
+    // ReplicationFactor defines the policy for log stream replication.
+    //
+    // +required
+    ReplicationFactor int32 `json:"replicationFactor,omitempty"`
+
+    // Limits defines the limits to be applied to log stream processing.
+    //
+    // +optional
+    Limits *LimitsSpec `json:"limits,omitempty"`
+
+    // Template defines the resource/limits/tolerations/nodeselectors per component
+    //
+    // +optional
+    Template *LokiTemplateSpec `json:"template,omitempty"`
+}
+```
+
+#### Optional: Global/Per-Tenant limits
+
+In general the `LimitsSpec` supports declaring limits for the ingestion and/or the query path individually. The settings can be applied either globally or per tenant. One exception to this rule is the ingestion limit `MaxGlobalStreamsPerTenant` which is applied on ingestion globally.
+
+```go
+// LimitsTemplateSpec defines the limits  applied at ingestion or query path.
+type LimitsTemplateSpec struct {
+    // IngestionLimits defines the limits applied on ingested log streams.
+    //
+    // +optional
+    IngestionLimits IngestionLimitSpec `json:"ingestion,omitempty"`
+
+    // QueryLimits defines the limit applied on querying log streams.
+    //
+    // +optional
+    QueryLimits QueryLimitSpec `json:"queries,omitempty"`
+}
+
+// LimitsSpec defines the spec for limits applied at ingestion or query
+// path across the cluster or per tenant.
+type LimitsSpec struct {
+
+    // Global defines the limits applied globally across the cluster.
+    //
+    // +optional
+    Global LimitsTemplateSpec `json:"global,omitempty"`
+
+    // Tenants defines the limits applied per tenant.
+    //
+    // +optional
+    Tenants map[string]LimitsSpec `json:"tenants,omitempty"`
+}
+```
+
+##### Query Limits
+
+The global/per-tenant query limits ensure sane default sizing of query result sets. In general users of [Loki HTTP API's](https://grafana.com/docs/loki/latest/api/) query capabilities can provide a `limit` and/or `start`, `end` parameters to control the query result set. However, if none given the limits below reflect sane defaults to keep the query path from unbound operations.
+
+In addition a sane default for `MaxChunksPerQuery` limits the required object storage operations and in turn provides sanity in object storage per-operation cost.
```

Comment:
i forget, did we talk about having something like a "profile" sort of sizing for the limits on the current api proposal or in a follow up? (to give a facade to the different knobs that loki currently exposes but may go away)

Reply:
No not really, we have not investigated or discussed this topic too much. There is always an implication that selecting a `Size: [1x.extra-small|1x.small|1x.medium]` will apply some sane defaults on the limits.